upcoming/ssh-server-in-go.md

```diff
@@ -0,0 +1,240 @@
+		_, stderr, err := com.ExecCmd("ssh-keygen", "-f", keyPath, "-t", "rsa", "-N", "")
+		if err != nil {
+			panic(fmt.Sprintf("Fail to generate private key: %v - %s", err, stderr))
+		}
+		log.Trace("New private key is generateed: %s", keyPath)
+	}
+
+	privateBytes, err := ioutil.ReadFile(keyPath)
+	if err != nil {
+		panic("Fail to load private key")
+	}
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		panic("Fail to parse private key")
+	}
+	config.AddHostKey(private)
+
+	go listen(config, port)
+}
+```
+
+This piece of code does three things:
+
+1. Setup a callback for validating public key from database.
+
+    Function `ssh.MarshalAuthorizedKey` will return a string format of user's public
+    key with a line break, so we want to remove that by calling `strings.TrimSpace`,
+    and then search in the database.
+
+    After search, if we return any kind of error, it will produce `Permission denied`
+    prompt on user side. If no error is returned, you can carry an instance of type
+    `*ssh.Permissions` to the corresponding request handler.
+
+    In this case, we need to set which key ID is this request corresponding to in `Extensions`.
+
+2. Create a private key when there is no one exists.
+
+    This is done by calling a command `ssh-keygen -f keypath -t rsa -N ""`.
+
+3. Load private key and start listening on given port.
+
+### Start listening and accepting new requests
+
+Like normal HTTP server, an SSH server needs to listen on a specific port as well.
+
+The pattern is very similar:
+
+```go
+func listen(config *ssh.ServerConfig, port int) {
+	listener, err := net.Listen("tcp", "0.0.0.0:"+com.ToStr(port))
+	if err != nil {
+		panic(err)
+	}
+	for {
+		// Once a ServerConfig has been configured, connections can be accepted.
+		conn, err := listener.Accept()
+		if err != nil {
+            // handle error
+			continue
+		}
+		// Before use, a handshake must be performed on the incoming net.Conn.
+		sConn, chans, reqs, err := ssh.NewServerConn(conn, config)
+		if err != nil {
+            // handle error
+			continue
+		}
+
+		// The incoming Request channel must be serviced.
+		go ssh.DiscardRequests(reqs)
+		go handleServerConn(sConn.Permissions.Extensions["key-id"], chans)
+	}
+}
+```
+
+1. Accept requests inside a infinite `for` loop.
+2. Preform handshakes for new SSH connections.
+3. Discard all irrelevant incoming request but serve the one you really need to care.
+
+    At this point, you can see we use `Extensions` to pass the user's public key ID
+    in the database.
+
+### Handle connections
+
+Finally, we're going to really serve the SSH requests.
+
+```go
+func handleServerConn(keyID string, chans <-chan ssh.NewChannel) {
+	for newChan := range chans {
+		if newChan.ChannelType() != "session" {
+			newChan.Reject(ssh.UnknownChannelType, "unknown channel type")
+			continue
+		}
+
+		ch, reqs, err := newChan.Accept()
+		if err != nil {
+            // handle error
```

Comment:
The indentation on this is spaces while the rest of the code is tabs. Can you change this to tabs. The difference causes display issues on the blog.